scheduler/launchd.go

```diff
@@ -0,0 +1,149 @@
+	launchdFileWasCreated bool, err error) {
+
+	if _, err := fileFunctions.OsStat(launchdAgentFilename); os.IsNotExist(err) {
+		// file does not exist, write out default launchd agent file
+		if err = fileFunctions.IoutilWriteFile(
+			launchdAgentFilename, []byte(LaunchdFileContents), 0644); err != nil {
+			return false,
+				fmt.Errorf("%s didn't exist and failed to create it: %v", launchdAgentFilename, err)
+		}
+	}
+
+	_, err = ld.runLaunchctl("load", launchdAgentFilename)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (ld *launchd) disable(fileFunctions fileFunctionsInterface, launchdAgentFilename string) (
+	launchdFileWasRemoved bool, err error) {
+
+	_, err = ld.runLaunchctl("unload", launchdAgentFilename)
+	if err != nil {
+		return false, err
+	}
+
+	if err := fileFunctions.OsRemove(launchdAgentFilename); err != nil {
+		return false, fmt.Errorf("failed to remove %s: %v", launchdAgentFilename, err)
+	}
+
+	return true, nil
+}
+
+func (ld *launchd) GetAgentFilename() (string, error) {
+	homeDirectory, err := homedir.Dir()
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(homeDirectory, LaunchAgentsSubdirectory, LaunchAgentFilename), nil
+}
+
+func (ld *launchd) runLaunchctl(verb string, filename string) (string, error) {
+	log.Printf("Running `%s %s %s`", launchctl, verb, filename)
+	cmd := exec.Command(launchctl, verb, filename)
+
+	out, err := cmd.CombinedOutput()
+
+	outString := string(out)
+	if err != nil {
+		return outString, err
+	}
+	if outString == filename+": No such file or directory\n" {
+		return outString, errCouldntFindLaunchdFile
+	}
+	return outString, nil
+}
+
+const launchctl string = "launchctl"
+
+// LaunchdFileContents is the agent file for running fk sync every 60 minutes
+const LaunchdFileContents = `<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+    <dict>
+        <key>Label</key>
+        <string>com.fluidkeys.sync</string>
+        <key>ProgramArguments</key>
+        <array>
+            <string>/usr/local/bin/fk</string>
+            <string>sync</string>
+            <string>--cron-output</string>
+        </array>
+        <key>StartInterval</key>
+        <integer>3600</integer>
+    </dict>
+</plist>
+`
+
+var (
+	// LaunchAgentsSubdirectory is the subdirectory for agent files, from the user's home directory
+	LaunchAgentsSubdirectory = path.Join("Library", "LaunchAgents")
+	// LaunchAgentFilename is the filename for the launchd agent
+	LaunchAgentFilename       = "com.fluidkeys.launchd.plist"
+	errCouldntFindLaunchdFile = errors.New("couldn't find launchd file")
+)
+
+// ErrModifyingLaunchd is a custom error. It should be used to wrap any errors called during
+// Enable or Disable to ensure the caller knows how to present the user with the steps they
+// can take to manually recify the situation.
+type ErrModifyingLaunchd struct {
```

Comment:
With crontab we know there's a very specific case we're looking for: where you call the `crontab` command to update the crontab, and due to permissions, it fails.

What's the equivalent here?

Is writing a `launchd` file any more likely to fail than writing, say, our `db.json` file? (We don't make special instructions for the user to manually edit their config file...)

Did you implement this just for symmetry or is there some other way launchd can fail that I haven't spotted yet?

Reply:
Good question. I was taking my lead from `cron`, so for symmetry. In hindsight, the only way it's going to fail is if it can't write that file to disk.

Reply:
*Can* `launchctl` fail? Does it have some permissions itself?